pkg/storage/stores/shipper/indexshipper/gatewayclient/gateway_client.go

```diff
@@ -122,47 +131,113 @@ func NewGatewayClient(cfg IndexGatewayClientConfig, r prometheus.Registerer, lim
+	sgClient.cfg.PoolConfig.HealthCheckIngesters = true
 
 	if sgClient.cfg.Mode == indexgateway.RingMode {
-		factory := func(addr string) (ring_client.PoolClient, error) {
-			igPool, err := NewIndexGatewayGRPCPool(addr, dialOpts)
+		sgClient.pool = clientpool.NewPool(sgClient.cfg.PoolConfig, sgClient.ring, factory, logger)
+	} else {
+		// Note we don't use clientpool.NewPool because we want to provide our own discovery function
+		poolCfg := ring_client.PoolConfig{
+			CheckInterval:      sgClient.cfg.PoolConfig.ClientCleanupPeriod,
+			HealthCheckEnabled: sgClient.cfg.PoolConfig.HealthCheckIngesters,
+			HealthCheckTimeout: sgClient.cfg.PoolConfig.RemoteTimeout,
+		}
+		clients := prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "loki",
+			Name:      "index_gateway_clients",
+			Help:      "The current number of index gateway clients.",
+		})
+		if r != nil {
+			err := r.Register(clients)
 			if err != nil {
-				return nil, errors.Wrap(err, "new index gateway grpc pool")
+				alreadyErr, ok := err.(prometheus.AlreadyRegisteredError)
+				if !ok {
+					return nil, err
+				}
+				clients = alreadyErr.ExistingCollector.(prometheus.Gauge)
 			}
-
-			return igPool, nil
 		}
+		//TODO(ewelch) we can't use metrics in the provider because of duplicate registration errors
+		dnsProvider := dns.NewProvider(logger, nil, dns.GolangResolverType)
+		sgClient.dnsProvider = dnsProvider
 
-		sgClient.pool = clientpool.NewPool(cfg.PoolConfig, sgClient.ring, factory, logger)
-	} else {
-		sgClient.conn, err = grpc.Dial(cfg.Address, dialOpts...)
-		if err != nil {
-			return nil, errors.Wrap(err, "index gateway grpc dial")
+		// Make an attempt to do one DNS lookup so we can start with addresses
+		sgClient.runDiscovery()
+		go sgClient.discoveryLoop()
+
+		discovery := func() ([]string, error) {
+			return dnsProvider.Addresses(), err
 		}
+		sgClient.pool = ring_client.NewPool("index gateway", poolCfg, discovery, factory, clients, logger)
+
+	}
 
-		sgClient.grpcClient = logproto.NewIndexGatewayClient(sgClient.conn)
+	// We have to start the pool service, it will handle removing stale clients in the background
+	err = services.StartAndAwaitRunning(context.Background(), sgClient.pool)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to start index gateway connection pool")
 	}
 
 	return sgClient, nil
 }
 
+func (s *GatewayClient) discoveryLoop() {
+	ticker := time.NewTicker(s.cfg.PoolConfig.ClientCleanupPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			s.runDiscovery()
+		case <-s.done:
+			return
+		}
+	}
+}
+
+func (s *GatewayClient) runDiscovery() {
+	ctx, cancel := context.WithTimeoutCause(context.Background(), 5*time.Second, ErrIndexGatewayDNSLookupTimeout)
+	defer cancel()
+	err := s.dnsProvider.Resolve(ctx, []string{s.cfg.Address})
+	if err != nil {
+		level.Error(s.logger).Log("msg", "failed to resolve index gateway address", "err", err)
+	}
+}
```

Comment:
The discovery loop should be encapsulated in the DNS service discovery that is passed to the `ring_client.NewPool()` function. Wdyt?

Reply:
yeah, agree with you.  I pulled it out and put it into the util directory.

Didn't really spend a ton of time to make it super generic but at least it may be easier for other folks to find now.